neon_iris/mq_connector.py

```diff
@@ -0,0 +1,141 @@
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+# OR PROFITS;  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE,  EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from time import sleep
+
+import pika.exceptions
+
+from asyncio import Event as AsyncEvent
+from threading import Event, Thread
+from neon_mq_connector.utils.client_utils import MQConnector
+from ovos_utils import LOG
+from pika.adapters.select_connection import SelectConnection
+from pika.channel import Channel
+
+
+class IrisConnector(MQConnector, Thread):
+    async_consumers_enabled = True
+
+    def __init__(self, *args, **kwargs):
+        vhost = kwargs.pop('vhost')
+        # TODO: service_name chosen for backwards-compat. and should be updated
+        #   to something more descriptive
+        kwargs['service_name'] = 'mq_handler'
+        Thread.__init__(self, daemon=True)
+        MQConnector.__init__(self, *args, **kwargs)
+        self.vhost = vhost
+        self._ready = AsyncEvent()
+        self._channel_closed = Event()
+        self._stopping = False
+
+        self._connection = self.init_connection()
+
+    def wait_for_connection(self):
+        LOG.info("Waiting for connection")
+        while not self._ready.is_set():
+            sleep(0.5)
+        LOG.info("Connected!")
+
+    @property
+    def connection(self) -> SelectConnection:
+        if self._connection is None:
+            self._connection = self.init_connection()
+        return self._connection
+
+    @property
+    def ready(self) -> bool:
+        return self._ready.is_set()
+
+    def run(self, *args, **kwargs):
+        MQConnector.run(self)
+        self._connection.ioloop.start()
+
+    def init_connection(self) -> SelectConnection:
+        return SelectConnection(
+            parameters=self.get_connection_params(self.vhost),
+            on_open_callback=self.on_connected,
+            on_open_error_callback=self.on_connection_fail,
+            on_close_callback=self.on_close)
+
+    def on_connected(self, _):
+        """Called when we are fully connected to RabbitMQ"""
+        LOG.info("MQ Connected")
+        self.connection.channel(on_open_callback=self.on_channel_open)
+
+    def on_connection_fail(self, *_, **__):
+        """ Called when connection to RabbitMQ fails"""
+        LOG.error(f"Failed to connect to MQ")
+        self._connection = None
+
+    def on_channel_open(self, new_channel: Channel):
+        """Called when our channel has opened"""
+        LOG.info(f"MQ Channel opened.")
+        new_channel.add_on_close_callback(self.on_channel_close)
+        self._ready.set()
+
+    def on_channel_close(self, *_, **__):
+        LOG.info(f"Channel closed.")
+        self._channel_closed.set()
+
+    def on_close(self, _, e):
```

Comment:
Do we know what _ is? Any reason we may want to document its type for future use?

Reply:
I believe it is the connection that is closed. I don't know of any reason we'd need the closed connection object, but I think I can add a type annotation without the IDE complaining about an unused parameter